rfc-123-github-actions-ci.md

```diff
@@ -0,0 +1,76 @@
+- GitHub Actions CI (hosted, paid through GDS' GitHub Enterprise Cloud subscription)
+
+## Proposal
+
+To switch to GitHub Actions CI for continuous integration (that is, branch and PR builds for tests/linting) as the preferred approach where we do not need the wider platform integration of Jenkins.
+
+This would involve working towards GOV.UK projects only using either GitHub Actions or Jenkins for CI, no projects should use Concourse, Travis or another tool for CI without reasons to justify an exception.
+
+This would be a gradual process, because we have many repos, but this decision would deem the usage of Travis CI as a legacy approach to CI for GOV.UK.
+
+### Examples of existing uses of GitHub Actions
+
+- [GOV.UK Coronavirus Business Volunteer Form][business-volunteer-form], a Rails app with a Postgres database.
+- [GOV.UK AWS linting checks][govuk-aws].
+- [GOV.UK Coronavirus Content YAML validation][govuk-coronavirus-content].
+
+### Pros
+
+- A simpler, more modern approach compared to Jenkins.
+- GDS as a whole pays for GitHub Enterprise Cloud, which entitles us to unlimited minutes on public repos (both Linux and macOS runners), and 50,000 minutes per month of Actions on private repos.
+- Supported by GitHub. We get Enterprise support via our GitHub Enterprise Cloud subscription. And we don't have to host any of our own infrastructure.
+- Well documented, easy to [get started][actions-get-started] building a new pipeline. Easier than Concourse for CI as PR builds are built-in to GitHub rather than needing to use [third-party code](https://github.com/telia-oss/github-pr-resource).
+- Doesn't require access to the VPN like our Jenkins does, or membership of any GitHub teams, in order to see logs for public repos.
+- External contributors can see their changes and why they failed, without us having to [push branches][merge-a-pull-request] and paste build logs from Jenkins. This fits more readily with our "code in the open" ethos, and saves on internal developer time.
+- Already in use across [some][business-volunteer-form] of [our][find-support-form] [production][vulnerable-people-form] applications.
+- As we go further down the path, we can make our internal GitHub Actions into libraries so we’re not duplicating “set up Rails apps” steps across 50 apps and can update them in one place - like we do with `alphagov/govuk-jenkinslib` now. Maybe `actions/alphagov`? (See [Homebrew/actions](https://github.com/Homebrew/actions) for an example.)
+- Getting rid of Jenkins would be a huge cost saving - our CI environment is in Carrenza and currently costs ~£6k a month + VAT.
+
+### Cons
+
+- We have less control over what software is on the runners. But most of the time we only need a database, a version of Ruby and some gems. And updating our Jenkins CI instances with new packages is a massive pain.
+- How do we do credentials management for things like pushing draft releases and Docker images to DockerHub?
```

Comment:
I think any CI process that uses secrets will have to run on Concourse to get all of the assurances we need.

Reply:
We should. @bevanloon @richardTowers Can we have a chat about this? It also relates to [this gem publishing comment](https://github.com/alphagov/rubocop-govuk/pull/41#issuecomment-618378751). (We're already doing gem publishing with GitHub Actions, and it's working well.) There's some detailed discussion on GitHub Actions secrets that I linked to in the above `rubocop-govuk` PR - that's worth a read.

Reply:
That said, can anyone think of any CI (ie linting and app tests - as opposed to *deployment processes*) that need secrets to run?

Reply:
I'm not aware of anything. However, if we don't allow secrets, do we have an extra consequence that gem releases need to happen <somewhere else>?

Reply:
I think it would be easier to make the "no secrets" decision for now - that still lets us do the majority of CI things, but as Ben says, we'll need to decide where to do gem releases (because we can't really do those in GitHub Actions without doing the secrets / IA / cyber dance).

Reply:
👍 - sounds like we'll be wanting to revert part of the demo for [rubocop-govuk](https://github.com/alphagov/rubocop-govuk/pull/45) and go with [some kind of restricted Jenkins build](https://github.com/alphagov/govuk-jenkinslib/pull/70) instead. That does seem like a slightly worse situation than what we had before, but I guess it only applies to a few repos - we only have a few gems.

Reply:
The pact contract testing process requires secrets to publish new pacts (used by Publishing API, GDS API Adapters and maybe Content Store).

I do think we'd be creating quite a pain point if we go down a no secrets route and would mean that we retrospectively need to undo some travis jobs we have that release npm modules. So I do think we should resolve it rather than going down "no secrets" as a simple option.

Reply:
Can we go forwards not backwards, please? I'd also argue that any gem publishing we do on Jenkins (if we must) should be done on the deploy Jenkinses rather than the CI Jenkinses - for separation of concerns.

Reply:
(Also my apologies, that previous comment came across a bit angry.)

Reply:
Maybe "no secrets" is the wrong heuristic. I think the gem ones are sensitive because there's a possible vector for compromising GOV.UK in there:

1) steal the rubygems secret from GitHub actions somehow
2) publish a malicious version of a gem that we use
3) wait for dependabot to update it
4) ?????
5) very bad things

Not to say that won't be an acceptable risk once we've spoken to cyber and IA (depends on how we feel about GitHub actions' security), but they definitely need to be in the loop.

The pact example is much less problematic, as it's not going to affect production code.

Also worth noting that I don't actually understand how we used to publish our gems, so I'm not sure what going forwards and backwards actually means...

Reply:
> Maybe "no secrets" is the wrong heuristic. I think the gem ones are sensitive because there's a possible vector for compromising GOV.UK in there:
> 
> 1. steal the rubygems secret from GitHub actions somehow
> 2. publish a malicious version of a gem that we use
> 3. wait for dependabot to update it
> 4. ?????
> 5. very bad things
> 
> Not to say that won't be an acceptable risk once we've spoken to cyber and IA (depends on how we feel about GitHub actions' security), but they definitely need to be in the loop.
> 
> The pact example is much less problematic, as it's not going to affect production code.
> 
> Also worth noting that I don't actually understand how we used to publish our gems, so I'm not sure what going forwards and backwards actually means...

I'd suggest that same vector applies for any gem and even our ones currently as there are already means that either our account or RubyGems is compromised. So that vector should be covered by always verifying what you're merging with dependabot (whether people always do is another matter)

I'd describe our risk as reputational, since were a token to be compromised then GOV.UK's name would be associated with it.

On something of a tangent the move to GitHub actions might also reduce security risks from a different angle as, for apps enabled, we would be building dependabot PRs on GitHubs infrastructure rather than our own.

Reply:
OK so I had a chat with Richard offline (thanks!) and we would like to reiterate that this RFC was only supposed to be about CI (branch builds for testing and linting). For *this case* GitHub Actions is fine. For gems and projects that aren't production code, GitHub Actions is fine for releasing too. For instance, `rubocop-govuk` is only a dev and test gem - and Kevin's `npm` pushes are fine as `html-to-govspeak` isn't a production service either, so it's fine to keep them as-is.

There may well be extensions to this RFC in future when we're at the point where IA/Cyber have done their assessments into GitHub Actions secrets security for *production* secrets, but that's the future. The *now* on this is GitHub Actions for CI builds, to save us a tonne of money and increase developer productivity.

Thank you. Now where's the "also send to channel" button on for GitHub threads...